### PR TITLE
Moved storage watcher into steth

### PR DESCRIPTION
**Pull request by sergeyWh1te**

### CODE REVIEW

File: ethereum-steth/src/services/withdrawals/Withdrawals.srv.ts [hunk 1]
```diff
@@ -805,53 +804,20 @@ export class WithdrawalsSrv {
...
-        `\tnot finalized: ${nonFinalizedStETH.dividedBy(ETH_DECIMALS).toFixed(4)}  ${notFinalized}  \n` +
-        `\tclaimed:       ${claimedStEth.dividedBy(ETH_DECIMALS).toFixed(4)} ${claimed} \n` +
-        `\tnot claimed:   ${nonClaimedStEth.dividedBy(ETH_DECIMALS).toFixed(4)}  ${notClaimed}\n` +
-        `\ttotal:         ${data.right.length} wr`,
+        `\tStEth:         ${stat.right.steth.toFixed(4)} \n` +
+        `\tfinalized:     ${stat.right.finalizedSteth.toFixed(4)} ${stat.right.finalizedRequests} \n` +
+        `\tnot finalized: ${stat.right.notFinalizedSteth.toFixed(4)}  ${stat.right.notfinalizedRequests}  \n` +
+        `\tclaimed:       ${stat.right.claimedSteth.toFixed(4)} ${stat.right.claimedRequests} \n` +
+        `\tnot claimed:   ${stat.right.notClaimedSteth.toFixed(4)}  ${stat.right.notClaimedRequests}\n` +
+        `\ttotal:         ${stat.right.total} wr`,
```
Comment (TheDZhon) [inline ethereum-steth/src/services/withdrawals/Withdrawals.srv.ts:820]: can't grasp the meaning of `total`

File: ethereum-steth/src/utils/constants.ts [hunk 2]
```diff
@@ -1,11 +1,20 @@
 import BigNumber from 'bignumber.js'
+import { StorageSlot } from '../entity/storage_slot'
 
 export type ERC20 = {
   decimals: number
   name: string
 }
 
 export type Address = {
+  NOR_ADDRESS: string
```
Comment (TheDZhon) [inline ethereum-steth/src/utils/constants.ts:10]: :warning:  Let's add SIMPLEDVT_ADDRESS and have the same storage slots watched as for NOR

File: ethereum-steth/src/utils/constants.ts [hunk 3]
```diff
@@ -70,9 +102,568 @@ export const Address: Address = {
...
+  {
+    id: 2,
+    contractAddress: LIDO_STETH_ADDRESS,
+    contactName: CONTACT_NAME_LIDO_STETH_TOKEN,
+    slotName: SLOT_LIDO_LOCATOR,
+    isAddress: false,
+    isArray: false,
+  },
+
+  // Node Operators Registry
```
Comment (TheDZhon) [inline ethereum-steth/src/utils/constants.ts:222]: :warning:  we should add SimpleDVT here too (it uses the same contract, but a different instance)

**Review by TheDZhon — CHANGES_REQUESTED** [review]
LGTM mostly! :+1: 

I'd suggest to:
- add SimpleDVT slots (they are the same as for NOR, but the different contract instance is used)
- adjust severity to `HIGH` since we use 🚨
- update README to reflect the up-to-date bot responsibilities (watching storage slots)
- improve readability of the tests
- don't use the `non-` prefix but stick to the already settled `un-` for `finalized`, `claimed`, etc.

File: ethereum-steth/src/clients/eth_provider.spec.ts [hunk 4]
```diff
@@ -214,6 +269,800 @@ describe('eth provider tests', () => {
       throw stBalance.left.message
     }
 
-    logger.info(stBalance.right.div(ETH_DECIMALS).toNumber())
+    expect(stBalance.right.div(ETH_DECIMALS).toNumber()).toEqual(13.292691238482098)
+  }, 120_000)
+
+  // Test storage watcher network methods
+  const blockHash = `0x0b99aebc4925ff127f1368b4aafff11dacc051a24247c0ee4159b735ca300d49`
+  const sloId = 777 // for test purposes id does not matter
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/clients/eth_provider.spec.ts:277]: done

File: ethereum-steth/src/clients/eth_provider.spec.ts [hunk 5]
```diff
@@ -1,21 +1,76 @@
...
   WithdrawalQueueERC721__factory,
 } from '../generated/typechain'
-import { ETHProvider } from './eth_provider'
-import { Metrics } from '../utils/metrics/metrics'
-import promClient from 'prom-client'
 import { HISTORY_BLOCK_OFFSET } from '../services/steth_operation/StethOperation.srv'
+import {
+  Address,
+  ETH_DECIMALS,
+  GATE_SEAL_DEFAULT_ADDRESS_BEFORE_26_APR_2024,
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/clients/eth_provider.spec.ts:19]: removed

File: ethereum-steth/src/clients/eth_provider.spec.ts [hunk 6]
```diff
@@ -214,6 +269,800 @@ describe('eth provider tests', () => {
...
+      address.NOR_ADDRESS,
+      sloId,
+      SLOT_NOR_STUCK_PENALTY_DELAY,
+      blockHash,
+    )
+    if (E.isLeft(data)) {
+      throw data.left.message
+    }
+
+    expect(data.right.value).toEqual('0x0000000000000000000000000000000000000000000000000000000000069780')
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/clients/eth_provider.spec.ts:338]: done

File: ethereum-steth/src/clients/eth_provider.spec.ts [hunk 7]
```diff
@@ -214,6 +269,800 @@ describe('eth provider tests', () => {
...
+      address.LIDO_INSURANCE_ADDRESS,
+      sloId,
+      SLOT_LIDO_INSURANCE_OWNER,
+      blockHash,
+    )
+    if (E.isLeft(data)) {
+      throw data.left.message
+    }
+
+    expect(data.right.value).toEqual('0x0000000000000000000000003e40d73eb977dc6a537af587d48316fee66e9c8c')
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/clients/eth_provider.spec.ts:1066]: done

File: ethereum-steth/src/entity/storage_slot.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+export type StorageSlot = {
+  id: number
+  contractAddress: string
+  contactName: string
+  slotName: string
+  slotAddress?: string
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/entity/storage_slot.ts:6]: Some data is fetched by name, some by address. You can find explanation in constants.ts

File: ethereum-steth/src/entity/withdrawal_request.ts [hunk 9]
```diff
@@ -3,12 +3,25 @@ import { WithdrawalQueueBase } from '../generated/typechain/WithdrawalQueueERC72
...
-  isClaimed: number
+  finalized: number
+  claimed: number
+}
+
+export type WithdrawalStat = {
+  finalizedSteth: number
+  notFinalizedSteth: number
+  claimedSteth: number
+  notClaimedSteth: number
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/entity/withdrawal_request.ts:18]: done

File: ethereum-steth/src/entity/withdrawal_request.ts [hunk 10]
```diff
@@ -3,12 +3,25 @@ import { WithdrawalQueueBase } from '../generated/typechain/WithdrawalQueueERC72
...
   timestamp: number
-  isFinalized: number
-  isClaimed: number
+  finalized: number
+  claimed: number
+}
+
+export type WithdrawalStat = {
+  finalizedSteth: number
+  notFinalizedSteth: number
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/entity/withdrawal_request.ts:16]: done

File: ethereum-steth/src/entity/withdrawal_request.ts [hunk 11]
```diff
@@ -3,12 +3,25 @@ import { WithdrawalQueueBase } from '../generated/typechain/WithdrawalQueueERC72
...
+  claimedSteth: number
+  notClaimedSteth: number
+  steth: number
+  total: number
+  finalizedRequests: number
+  notfinalizedRequests: number
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/entity/withdrawal_request.ts:22]: done

File: ethereum-steth/src/entity/withdrawal_request.ts [hunk 12]
```diff
@@ -3,12 +3,25 @@ import { WithdrawalQueueBase } from '../generated/typechain/WithdrawalQueueERC72
...
+  finalized: number
+  claimed: number
+}
+
+export type WithdrawalStat = {
+  finalizedSteth: number
+  notFinalizedSteth: number
+  claimedSteth: number
+  notClaimedSteth: number
+  steth: number
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/entity/withdrawal_request.ts:19]: stethAmount

File: ethereum-steth/src/handlers/init.handler.ts [hunk 13]
```diff
@@ -62,9 +67,12 @@ export class InitHandler {
       ]
       metadata.agents = '[' + agents.toString() + ']'
 
-      const withdrawalsSrvErr = await this.WithdrawalsSrv.initialize(this.latestBlockNumber)
+      const [withdrawalsSrvErr, storageWatcherErr] = await Promise.all([
+        this.WithdrawalsSrv.initialize(this.latestBlock.number),
+        this.storageWatcher.initialize(this.latestBlock.hash),
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/handlers/init.handler.ts:72]: fixed

File: ethereum-steth/src/handlers/tx.handler.ts [hunk 14]
```diff
@@ -45,6 +45,7 @@ export class TxHandler {
 
       const findings: Finding[] = []
 
+      // TODO to group logs by needed contract addresses.
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/handlers/tx.handler.ts:48]: Leave it I'll fix it next iteration

File: ethereum-steth/src/main.ts [hunk 15]
```diff
@@ -129,10 +130,12 @@ const main = async () => {
     address.LIDO_STETH_ADDRESS,
   )
 
+  const storageWatcherSrv = new StorageWatcherSrv(logger, STORAGE_SLOTS, ethClient)
+
   try {
     await dbClient.migrate.latest()
 
-    const sql = fs.readFileSync('./src/db/withdrawal_requests_17_07_24.sql', 'utf8')
+    const sql = fs.readFileSync('./src/db/withdrawal_requests_01_08_24.sql', 'utf8')
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/main.ts:138]: I really could not understand: Cause in latest commit withdrawal_requests_01_08_24

File: ethereum-steth/src/services/withdrawals/Withdrawals.srv.ts [hunk 16]
```diff
@@ -395,7 +394,7 @@ export class WithdrawalsSrv {
           // if we are in turbo mode and unfinalized queue is not finalized for 5 days
           // and alert hasn't been sent for 1 day
           const f: Finding = new Finding()
-          f.setName(`⚠️ Withdrawals: unfinalized queue wait time is more than ${FIVE_DAYS / ONE_DAY} days`)
+          f.setName(`⚠️ Withdrawals: unfinalized queue wait time is more than ${FIVE_DAYS} days`)
```
[the remark this answers is not in the document]
Comment (sergeyWh1te) [reply ethereum-steth/src/services/withdrawals/Withdrawals.srv.ts:397]: Restored back, thx!

**Review by TheDZhon — APPROVED** [review]
Thank you! :+1:

**Comment by TheDZhon** [comment]
Let's update the README file before merging this

**Review by TheDZhon — APPROVED** [review]

**Pull request merged**